Shared/AzureFunctions/Get-GraphAccessToken.ps1

```diff
@@ -0,0 +1,204 @@
+﻿# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+. $PSScriptRoot\..\ScriptUpdateFunctions\Invoke-WebRequestWithProxyDetection.ps1
+
+function Get-GraphAccessToken {
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory = $false)]
+        [string]$AzureADEndpoint = "https://login.microsoftonline.com",
+
+        [Parameter(Mandatory = $false)]
+        [string]$GraphApiUrl = "https://graph.microsoft.com",
+
+        [Parameter(Mandatory = $false)]
+        [string]$Scope = "$($GraphApiUrl)//AuditLog.Read.All Directory.AccessAsUser.All email openid profile"
+    )
+
+    <#
+        This function is used to get an access token for the Azure Graph API by using the OAuth 2.0 authorization code flow
+        with PKCE (Proof Key for Code Exchange). The OAuth 2.0 authorization code grant type, or auth code flow,
+        enables a client application to obtain authorized access to protected resources like web APIs.
+        The auth code flow requires a user-agent that supports redirection from the authorization server
+        (the Microsoft identity platform) back to your application.
+
+        More information about the auth code flow with PKCE can be found here:
+        https://learn.microsoft.com/azure/active-directory/develop/v2-oauth2-auth-code-flow#protocol-details
+    #>
+
+    begin {
+        Write-Verbose "Calling $($MyInvocation.MyCommand)"
+        function StartLocalListener {
+            param(
+                [int]$Port = 8004,
+                [int]$TimeoutSeconds = 120
+            )
+
+            $url = $null
+            $authCompletedWording = "Authentication complete. You can return to the application. Feel free to close this browser tab."
+            $stopWatch = [system.diagnostics.stopwatch]::StartNew()
+            $listener = New-Object "Net.HttpListener"
+            $listener.Prefixes.add("http://localhost:$($Port)/")
+            try {
+                $listener.Start()
+                Start-Sleep -Seconds 2
+                while (($listener.IsListening) -and
+                    ($stopWatch.Elapsed.Seconds -le $TimeoutSeconds)) {
+                    $context = $listener.GetContext()
```

Comment:
This line makes us hang until a response is available. For example, if you simply close the browser window, the script is now stuck and you can't ctrl-C out of it. It would be better to use GetContextAsync() in a loop like ExTRA.ps1 does. Refer to that script for the full example.

```powershell
    while ($httpListener.IsListening) {
        $task = $httpListener.GetContextAsync()
        while (-not $task.AsyncWaitHandle.WaitOne(100)) {
            Start-Sleep -Milliseconds 100
        }
```

Reply:
Good catch! Will change it as requested

Reply:
Done